csharp/src/Ice/InputStream.cs

```diff
@@ -1070,62 +1037,101 @@ public void SkipSize()
-                if (bytesRead < data.Length)
-                {
-                    throw new InvalidDataException("invalid endpoint encapsulation size while reading opaque endpoint");
-                }
-                result = new OpaqueEndpoint(type, _encoding.Value, data);
+                endpoint = new OpaqueEndpoint(type, encoding, _buffer.Slice(_pos, size - 6).ToArray());
+                _pos += size - 6;
             }
 
-            if (_limit.Value - _pos != 0)
+            return endpoint;
+        }
+
+        private static (Encoding Encoding, int Size) ReadEncapsulationHeader(ReadOnlySpan<byte> buffer,
+                                                                             Encoding encoding)
+        {
+            Debug.Assert(encoding == Encoding.V1_1); // temporary
+
+            // With the 1.1 encoding, the encapsulation size is encoded on a 4-bytes int and not on a variable-length
+            // size, for ease of marshaling.
+            if (buffer.Length < 6)
             {
-                throw new InvalidDataException(
-                    $"there are {_limit.Value - _pos} bytes remaining in endpoint encapsulation");
+                throw new InvalidDataException($"encapsulation buffer has only {buffer.Length} bytes");
             }
-
-            // Exceptions when reading InputStream are considered fatal to the InputStream so no need to restore
-            // _limit or _encoding unless we succeed.
-            _limit = oldLimit;
-            _encoding = oldEncoding;
-            return result;
+            int size = ReadInt(buffer);
+            if (size < 6)
+            {
+                throw new InvalidDataException($"encapsulation has only {size} bytes");
+            }
+            if (size - 4 > buffer.Length)
+            {
+                throw new InvalidDataException("the encapsulation's size extends beyond the end of the buffer");
+            }
+            return (new Encoding(buffer[4], buffer[5]), size);
         }
 
-        internal (Encoding Encoding, int Size) ReadEncapsulationHeader()
+        private InputStream(Communicator communicator, ArraySegment<byte> buffer, bool startEncaps, int pos)
         {
-            // With the 1.1 encoding, the encapsulation size is encoded on a 4-bytes int and
-            // not on a variable-length size, for ease of marshaling.
-            int sz = ReadInt();
-            if (sz < 6)
+            Debug.Assert(pos == 0 || !startEncaps); // while pos is still there, it's 0 when startEncaps is true
+            Communicator = communicator;
+
+            if (startEncaps)
             {
-                throw new InvalidDataException($"encapsulation has only {sz} bytes");
+                _pos = 0;
+                int size;
+                (Encoding, size) = ReadEncapsulationHeader(buffer, Encoding.V1_1);
+                Encoding.CheckSupported();
+                _buffer = buffer.Slice(6, size - 6);
```

Comment:
Will be good to add a comment about why we Slice the buffer here. Later with ice2 we will also need to restore the buffer to read the data after the encapsulation.

Reply:
Added comment.
For the 2.0 encoding & ice2 protocol, I anticipate we'll read the context "footer" together with the header, so we'll still slice to the encapsulation.